buildSrc/subprojects/parser/src/main/kotlin/org/apache/calcite/buildtools/parser/DialectGenerateTask.kt

```diff
@@ -0,0 +1,210 @@
+        for (f in files) {
+            if (f.isFile && f.extension == "ftl") {
+                processFile(f, functionMap)
+            }
+            if (directories.isNotEmpty() && f.name == nextDirectory) {
+                directories.poll()
+                traverse(directories, f, functionMap)
+            }
+        }
+    }
+
+    /**
+     * Extracts the functions from the given file into functionMap. Parses
+     * functions of the form:
+     * <return_type> <name>(<args>) :
+     * {
+     *     <properties>
+     * }
+     * {
+     *     <body>
+     * }
+     * @param file The file to process
+     * @param functionMap The map to which the parsing functions will be added to
+     */
+    private fun processFile(file: File, functionMap: MutableMap<String, String>) {
+        var fileText = file.readText(Charsets.UTF_8)
+        val typeAndName = "\\w+\\s+\\w+"
+        val declarationPattern =
+            Regex("(%s\\s*\\(\\s*(%s\\s*(\\,\\s*%s\\s*)*)?\\)\\s*\\:\n)"
+                    .format(typeAndName, typeAndName, typeAndName))
+        val delims = "(\\s|\n|\"|(//)|(/\\*)|(\\*/))"
+        // Uses Lookahead and Lookbehind to ensure that the delims are added as tokens.
+        val splitRegex = Regex("((?<=%s)|(?=%s))".format(delims, delims))
+        val declarationMatches = declarationPattern.findAll(fileText)
+        for (m in declarationMatches) {
+            val functionDeclaration = m.value
+            val functionName = getFunctionName(functionDeclaration)
+            val functionBuilder = StringBuilder(functionDeclaration)
+            val declarationIndex = fileText.indexOf(functionDeclaration)
+            fileText = fileText.substring(declarationIndex + functionDeclaration.length)
+            val tokens: Queue<String> = LinkedList(splitRegex.split(fileText))
+            processCurlyBlock(functionBuilder, tokens)
+            processCurlyBlock(functionBuilder, tokens)
+            functionMap.put(functionName, functionBuilder.toString())
+        }
+    }
+
+    /**
+     * Parses a block of text surrounded by curly braces. The function keeps track
+     * of the number of curly braces that have yet to be closed. Once this counter
+     * reaches 0 the block has been fully parsed. The function also ensures that
+     * curly braces within single-line comments, multi-line comments, and strings
+     * are not counted for this.
+     *
+     * @param functionBuilder The builder unto which the tokens get added to once parsed
+     * @param tokens The tokens to be parsed
+     */
+    private fun processCurlyBlock(functionBuilder: StringBuilder, tokens: Queue<String>) {
+        var curlyCounter = 0
+        var insideString = false
+        var insideSingleLineComment = false
+        var insideMultiLineComment = false
+        var validBrace = true
+        while (tokens.isNotEmpty()) {
+            val token = tokens.poll()
+            functionBuilder.append(token)
```

Comment:
Shouldn't the token be appended after all the checks are cleared instead of before?

Reply:
It is added before since there is no case in which we want to not add a token to the builder. The checks are only there to check if we want to treat a curly brace as an actual curly brace.

Reply:
Great catch, Lisa! Yes, I think this would be undesirable if the first token is, say, a newline.

However, I think the main assumption here is that the first character will be an open curly brace. Although if we're making that assumption, it would be nice to codify it with some sort of an assert.

Reply:
Does this function only get passed legitimate curly blocks?

Reply:
This opens an interesting situation: do we assume that all of the functions in parserImpls.ftl are valid? If not, I can add some checks to ensure that it follows the assumed pattern.

Reply:
Currently the if block that checks for new line allows for an arbitrary amount of newlines before the first curly brace as I wasn't sure if that was a valid format or not.

Reply:
I'm a little confused what this function does it if puts every token into the function builder regardless of any of the checks. Is it supposed to leave out any information after the last closing brace?

Will this be a valid input to the function?

> {
>  info
> }
> {
>   second info
> }

Reply:
Yes, it would stop after the first closing brace. Currently the function is being called twice inside the processFile function in order to process both curly blocks.

Reply:
ah that makes sense